picard/ui/coverartbox.py

```diff
@@ -94,51 +95,119 @@ def __eq__(self, other):
-                pixmap.loadFromData(self.data.data)
-            if not pixmap.isNull():
-                offx, offy, w, h = (1, 1, 121, 121)
-                cover = QtGui.QPixmap(self.shadow)
-                pixmap = pixmap.scaled(w, h, QtCore.Qt.KeepAspectRatio, QtCore.Qt.SmoothTransformation)
-                painter = QtGui.QPainter(cover)
-                bgcolor = QtGui.QColor.fromRgb(0, 0, 0, 128)
-                painter.fillRect(QtCore.QRectF(offx, offy, w, h), bgcolor)
-                x = offx + (w - pixmap.width()) / 2
-                y = offy + (h - pixmap.height()) / 2
-                painter.drawPixmap(x, y, pixmap)
+                pixmap.loadFromData(self.data[0].data)
+                pixmap = self.decorate_cover(pixmap)
+            else:
+                limited = len(self.data) > 4
+                if limited:
+                    data_to_paint = data[:3]
+                    offset = displacements * len(data_to_paint)
+                else:
+                    data_to_paint = data
+                    offset = displacements * (len(data_to_paint) - 1)
+                stack_width, stack_height = (w + offset, h + offset)
+                pixmap = QtGui.QPixmap(stack_width, stack_height)
+                bgcolor = self.palette().color(QtGui.QPalette.Window)
+                painter = QtGui.QPainter(pixmap)
+                painter.fillRect(QtCore.QRectF(0, 0, stack_width, stack_height), bgcolor)
+                cx = stack_width - w / 2
+                cy = h / 2
+                if limited:
+                    x, y = (cx - self.shadow.width() / 2, cy - self.shadow.height() / 2)
+                    for i in range(3):
+                        painter.drawPixmap(x, y, self.shadow)
+                        x -= displacements / 3
+                        y += displacements / 3
+                    cx -= displacements
+                    cy += displacements
+                else:
+                    cx = stack_width - w / 2
+                    cy = h / 2
+                for image in reversed(data_to_paint):
+                    if isinstance(image, QtGui.QPixmap):
+                        thumb = image
+                    else:
+                        thumb = QtGui.QPixmap()
+                        thumb.loadFromData(image.data)
+                    thumb = self.decorate_cover(thumb)  # QtGui.QColor("darkgoldenrod")
+                    x, y = (cx - thumb.width() / 2, cy - thumb.height() / 2)
+                    painter.drawPixmap(x, y, thumb)
+                    cx -= displacements
+                    cy += displacements
+                if not has_common_images:
+                    color = QtGui.QColor("darkgoldenrod")
+                    border_length = 10
+                    for k in range(border_length):
+                        color.setAlpha(255 - k * 255 / border_length)
+                        painter.setPen(color)
+                        painter.drawLine(x, y - k - 1, x + 121 + k + 1, y - k - 1)
```

Comment:
AFAIK, there isn't any method in Qt for that. I think there may be methods in kde frameworks in some style library like breeze, but that's definitely of no use here. About high dpi scaling, that's "not supported" on Qt4, Qt5 brings support for hidpi devices from Qt 5.6.0, but that's definitely something to worry about with picard 2.0 . In the meantime, since the coverartbox is fixed to use 128x128 covers, that means users with hidpi devices will see small covers and small cover stacks (which Qt5 will scale automatically)